server/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java

```diff
@@ -55,6 +55,11 @@ default void afterIndexShardStarted(IndexShard indexShard) {}
      */
     default void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {}
 
+    /**
+     * Called after the index shard has been marked closed. It could still be waiting for the async close of the engine.
```

Comment:
nit Might be good to extend the comment to say that afterIndexShardClosed may be called earlier (due to asynchronicity in rare cases).

Reply:
Sure, done in 1217103fe087580889fe69e00307670dd4d56239